model/modeldecoder/generator/generator.go

```diff
@@ -71,17 +74,18 @@ func NewGenerator(importPath string, pkg string, typPath string,
 	if err != nil {
 		return nil, err
 	}
+	imports := make(map[string]struct{})
 	g := Generator{
-		pkgName:           loaded.Types.Name(),
-		structTypes:       structTypes,
-		rootObjs:          make([]structType, 0, len(root)),
-		processedTypes:    make(map[string]struct{}),
-		imports:           make(map[string]struct{}),
-		nullableString:    fmt.Sprintf("%s.String", typPath),
-		nullableInt:       fmt.Sprintf("%s.Int", typPath),
-		nullableFloat64:   fmt.Sprintf("%s.Float64", typPath),
-		nullableBool:      fmt.Sprintf("%s.Bool", typPath),
-		nullableInterface: fmt.Sprintf("%s.Interface", typPath),
+		pkgName:        loaded.Types.Name(),
+		structTypes:    structTypes,
+		rootObjs:       make([]structType, 0, len(root)),
+		processedTypes: make(map[string]struct{}),
+		imports:        imports,
+		validationGenerators: map[string]validationGenerator{
+			"struct": newTstruct(imports, typPath),
+			"slice":  newTslice(imports),
+			"map":    newTmap(imports),
+		},
```

Comment:
I'm finding this harder to follow now. Here's some suggestions that I think will simplify things. Feel free to defer until later; splitting apart the big function is a step in the right direction anyway.

Essentially I think it would be simpler if we didn't use an interface, but instead used static calls to type-specific generator functions. e.g. if it's a struct, call the struct validation generator directly, as opposed to looking up a generic interface in a map and calling it indirectly.

The first part would require lifting the nullable type handling out of tstruct. In NewGenerator, you could pull out references to the types like this:

```go
typPackage := loaded.Imports[typPath].Types
nullableString := typPackage.Scope().Lookup("String").Type()
...
```

(loadPackage needs to be modified to pass `packages.NeedImports` for this to work)

Then, in `generator.generateValidation`, you can then compare the type to those nullable types and call their validation functions directly, and otherwise use the type's shape to call one of the other tstruct/tmap/tslice validation functions directly.

The second part is to do with "imports". Instead of having stateful generator objects with references to an imports map, we could pass "imports" into the validation generator functions. Another would be to remove that altogether: add all imports, and use https://pkg.go.dev/golang.org/x/tools/imports#Process to remove any unused ones after generation. I'd go the latter, as it would simplify some of the generator logic.

In the end, `generator.generateValidation` would look more like this:

```go
var validation validationGenerator                                                                 
for i := 0; i < len(structTyp.fields); i++ {                                                       
        var custom bool                                                                            
        f := structTyp.fields[i]                                                                   
        switch f.Type() {                                                                          
        case g.nullableString:                                                                     
                validation = generateNullableStringValidation                                                              
        case g.nullableInt:                                                                        
                validation = generateNullableIntValidation                                                                 
        case g.nullableFloat64:                                                                    
                validation = generateNullableFloat64Validation                                                             
        case g.nullableInterface:                                                                  
                validation = generateNullableInterfaceValidation                                                           
        default:                                                                                   
                switch t := f.Type().Underlying().(type) {                                         
                case *types.Slice:                                                                 
                        validation = generateSliceValidation
                        _, custom = g.customStruct(t.Elem())                                       
                case *types.Map:                                                                   
                        validation = generateMapValidation                                                         
                        _, custom = g.customStruct(t.Elem())                                       
                case *types.Struct:                                                                
                        validation = generateStructValidation                                                      
                        _, custom = g.customStruct(f.Type())                                       
                default:                                                                           
                        return errors.Wrap(fmt.Errorf("unhandled type %T", t), flattenName(key, f))
                }                                                                                  
        }                                                                                          
        if err := validation.validation(&g.buf, structTyp.fields, f, custom); err != nil {         
                return errors.Wrap(err, flattenName(key, f))                                       
        }                                                                                          
}                                                                                                  
```

Reply:
Thanks for the great feedback, I have the code adapted accordingly.